subtests/docker_cli/events/events.py

```diff
@@ -98,42 +55,88 @@ def is_dupe_event(needle, haystack):
+        'operation':  mobj.group('operation'),
+        'source':     None,
+    }
+    # TODO: (maybe): split out components of the parenthesized list.
+    # If so, keep in mind that you can't just split on commas (because
+    # of "Red Hat, Inc.") and that the fields are output in unpredictable
+    # order: even two consecutive event lines will have different ordering.
+    source_re = re.compile(r'(^|\s)image=(?P<image>\S+)(,|$)')
+    mobj2 = source_re.search(mobj.group('rest'))
+    if mobj2 is not None:
+        details['source'] = mobj2.group('image')
+    return details
+
+
+def parse_event_docker_109(line):
+    """
+    Try to parse input as a docker < 1.10 event
+    """
+    # eg <timestamp> <sha> (from <source>) start
+    event_re = re.compile(r'^(?P<timestamp>{timestamp})'
+                          r'\s+(?P<identifier>{cid}|{fqin}):'
+                          r'(\s+\(from (?P<source>{source})\))?'
+                          r'\s+(?P<operation>{operation})'.format(**regexes))
+    mobj = event_re.match(line)
+    if mobj is not None:
+        return {
+            'datetime':   DockerTime(mobj.group('timestamp')),
+            'identifier': mobj.group('identifier'),
+            'source':     mobj.group('source'),
+            'operation':  mobj.group('operation'),
+        }
+    return None
+
+
 def parse_event(line):
     """
-    Return tuple(CID/FQIN, {DETAILS}) from parsing line
+    Return {DETAILS} from parsing line
 
     :param line: String-like containing a single event line
-    :returns: (CID/FQIN, {DETAILS}) from parsing line or None if unparseable
+    :returns: {DETAILS} from parsing line or None if unparseable
     """
-    identifier = event_cid(line)
-    if identifier is None:
-        identifier = event_fqin(line)
-    details = event_details(line)
-    if identifier is None or details['datetime'] is None:
-        return None  # unparseable line
-    else:
-        return (identifier, details)
+    details = parse_event_docker_110(line)
+    if details is None:
+        details = parse_event_docker_109(line)
+    return details
 
 
 def parse_events(lines, slop=None):
     """
     Return list of tuples for valid lines returned by parse_events()
 
     :param lines: String containing events, one per line
-    :param slop: number of unparseable lines to tollerate, None/- to disable
+    :param slop: number of unparseable lines to tolerate, None/- to disable
     :returns: List of tuple(CID, {DETAILS}) as returned from parse_events()
     """
     sloppy = []
     result = []
     n_lines = 0
     for line in lines.splitlines():
         n_lines += 1
-        if len(line) < 64:   # len of a cid
-            sloppy.append(line)
-            continue
         cid_details = parse_event(line)
         if cid_details is not None:
-            result.append(cid_details)
+            result.append((cid_details['identifier'], cid_details))
         else:
             sloppy.append(line)
```

Comment:
If you like, it might be okay to get rid of the sloppy support.  There was a time ages ago when it was actually necessary to get a passing test :scream: